File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax/CConst.cs
using System;
using System.Runtime.CompilerServices;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
{
	internal abstract class CConst
	{
		protected readonly ERRID _errid;

		protected readonly object[] _diagnosticArguments;

		public abstract SpecialType SpecialType { get; }

		public abstract object ValueAsObject { get; }

		public bool IsBad => SpecialType == SpecialType.None;

		public bool IsBooleanTrue
		{
			get
			{
				if (IsBad)
				{
					return false;
				}
				if (this is CConst<bool> cConst)
				{
					return cConst.Value;
				}
				return false;
			}
		}

		public ERRID ErrorId => _errid;

		public object[] ErrorArgs => _diagnosticArguments ?? Array.Empty<object>();

		public CConst()
		{
		}

		public CConst(ERRID id, params object[] diagnosticArguments)
		{
			_errid = id;
			_diagnosticArguments = diagnosticArguments;
		}

		public abstract CConst WithError(ERRID id);

		internal static CConst CreateChecked(object value)
		{
			return TryCreate(RuntimeHelpers.GetObjectValue(value));
		}

		internal static CConst TryCreate(object value)
		{
			if (value == null)
			{
				return CreateNothing();
			}
			return SpecialTypeExtensions.FromRuntimeTypeOfLiteralValue(RuntimeHelpers.GetObjectValue(value)) switch
			{
				SpecialType.System_Boolean => Create(Convert.ToBoolean(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_Byte => Create(Convert.ToByte(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_Char => Create(Convert.ToChar(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_DateTime => Create(Convert.ToDateTime(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_Decimal => Create(Convert.ToDecimal(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_Double => Create(Convert.ToDouble(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_Int16 => Create(Convert.ToInt16(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_Int32 => Create(Convert.ToInt32(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_Int64 => Create(Convert.ToInt64(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_SByte => Create(Convert.ToSByte(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_Single => Create(Convert.ToSingle(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_String => Create(Convert.ToString(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_UInt16 => Create(Convert.ToUInt16(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_UInt32 => Create(Convert.ToUInt32(RuntimeHelpers.GetObjectValue(value))), 
				SpecialType.System_UInt64 => Create(Convert.ToUInt64(RuntimeHelpers.GetObjectValue(value))), 
				_ => null, 
			};
		}

		internal static CConst<object> CreateNothing()
		{
			return new CConst<object>(null, SpecialType.System_Object);
		}

		internal static CConst<bool> Create(bool value)
		{
			return new CConst<bool>(value, SpecialType.System_Boolean);
		}

		internal static CConst<byte> Create(byte value)
		{
			return new CConst<byte>(value, SpecialType.System_Byte);
		}

		internal static CConst<sbyte> Create(sbyte value)
		{
			return new CConst<sbyte>(value, SpecialType.System_SByte);
		}

		internal static CConst<char> Create(char value)
		{
			return new CConst<char>(value, SpecialType.System_Char);
		}

		internal static CConst<short> Create(short value)
		{
			return new CConst<short>(value, SpecialType.System_Int16);
		}

		internal static CConst<ushort> Create(ushort value)
		{
			return new CConst<ushort>(value, SpecialType.System_UInt16);
		}

		internal static CConst<int> Create(int value)
		{
			return new CConst<int>(value, SpecialType.System_Int32);
		}

		internal static CConst<uint> Create(uint value)
		{
			return new CConst<uint>(value, SpecialType.System_UInt32);
		}

		internal static CConst<long> Create(long value)
		{
			return new CConst<long>(value, SpecialType.System_Int64);
		}

		internal static CConst<ulong> Create(ulong value)
		{
			return new CConst<ulong>(value, SpecialType.System_UInt64);
		}

		internal static CConst<decimal> Create(decimal value)
		{
			return new CConst<decimal>(value, SpecialType.System_Decimal);
		}

		internal static CConst<string> Create(string value)
		{
			return new CConst<string>(value, SpecialType.System_String);
		}

		internal static CConst<float> Create(float value)
		{
			return new CConst<float>(value, SpecialType.System_Single);
		}

		internal static CConst<double> Create(double value)
		{
			return new CConst<double>(value, SpecialType.System_Double);
		}

		internal static CConst<DateTime> Create(DateTime value)
		{
			return new CConst<DateTime>(value, SpecialType.System_DateTime);
		}
	}
	internal class CConst<T> : CConst
	{
		private readonly SpecialType _specialType;

		private readonly T _value;

		public override SpecialType SpecialType => _specialType;

		public override object ValueAsObject => _value;

		public T Value => _value;

		internal CConst(T value, SpecialType specialType)
		{
			_value = value;
			_specialType = specialType;
		}

		private CConst(T value, SpecialType specialType, ERRID id)
			: base(id)
		{
			_value = value;
			_specialType = specialType;
		}

		public override CConst WithError(ERRID id)
		{
			return new CConst<T>(_value, _specialType, id);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/InternalSyntax/CConst.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports System
Imports System.Runtime.CompilerServices

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
	Friend MustInherit Class CConst
		Protected ReadOnly _errid As ERRID

		Protected ReadOnly _diagnosticArguments As Object()

		Public ReadOnly Property ErrorArgs As Object()
			Get
				Return If(Me._diagnosticArguments, Array.Empty(Of Object)())
			End Get
		End Property

		Public ReadOnly Property ErrorId As ERRID
			Get
				Return Me._errid
			End Get
		End Property

		Public ReadOnly Property IsBad As Boolean
			Get
				Return Me.SpecialType = Microsoft.CodeAnalysis.SpecialType.None
			End Get
		End Property

		Public ReadOnly Property IsBooleanTrue As Boolean
			Get
				Dim flag As Boolean
				If (Not Me.IsBad) Then
					Dim cConst As CConst(Of Boolean) = TryCast(Me, CConst(Of Boolean))
					flag = If(cConst Is Nothing, False, cConst.Value)
				Else
					flag = False
				End If
				Return flag
			End Get
		End Property

		Public MustOverride ReadOnly Property SpecialType As Microsoft.CodeAnalysis.SpecialType

		Public MustOverride ReadOnly Property ValueAsObject As Object

		Public Sub New()
			MyBase.New()
		End Sub

		Public Sub New(ByVal id As ERRID, ByVal ParamArray diagnosticArguments As Object())
			MyBase.New()
			Me._errid = id
			Me._diagnosticArguments = diagnosticArguments
		End Sub

		Friend Shared Function Create(ByVal value As Boolean) As CConst(Of Boolean)
			Return New CConst(Of Boolean)(value, Microsoft.CodeAnalysis.SpecialType.System_Boolean)
		End Function

		Friend Shared Function Create(ByVal value As Byte) As CConst(Of Byte)
			Return New CConst(Of Byte)(value, Microsoft.CodeAnalysis.SpecialType.System_Byte)
		End Function

		Friend Shared Function Create(ByVal value As SByte) As CConst(Of SByte)
			Return New CConst(Of SByte)(value, Microsoft.CodeAnalysis.SpecialType.System_SByte)
		End Function

		Friend Shared Function Create(ByVal value As Char) As CConst(Of Char)
			Return New CConst(Of Char)(value, Microsoft.CodeAnalysis.SpecialType.System_Char)
		End Function

		Friend Shared Function Create(ByVal value As Short) As CConst(Of Short)
			Return New CConst(Of Short)(value, Microsoft.CodeAnalysis.SpecialType.System_Int16)
		End Function

		Friend Shared Function Create(ByVal value As UShort) As CConst(Of UShort)
			Return New CConst(Of UShort)(value, Microsoft.CodeAnalysis.SpecialType.System_UInt16)
		End Function

		Friend Shared Function Create(ByVal value As Integer) As CConst(Of Integer)
			Return New CConst(Of Integer)(value, Microsoft.CodeAnalysis.SpecialType.System_Int32)
		End Function

		Friend Shared Function Create(ByVal value As UInteger) As CConst(Of UInteger)
			Return New CConst(Of UInteger)(value, Microsoft.CodeAnalysis.SpecialType.System_UInt32)
		End Function

		Friend Shared Function Create(ByVal value As Long) As CConst(Of Long)
			Return New CConst(Of Long)(value, Microsoft.CodeAnalysis.SpecialType.System_Int64)
		End Function

		Friend Shared Function Create(ByVal value As ULong) As CConst(Of ULong)
			Return New CConst(Of ULong)(value, Microsoft.CodeAnalysis.SpecialType.System_UInt64)
		End Function

		Friend Shared Function Create(ByVal value As [Decimal]) As CConst(Of [Decimal])
			Return New CConst(Of [Decimal])(value, Microsoft.CodeAnalysis.SpecialType.System_Decimal)
		End Function

		Friend Shared Function Create(ByVal value As String) As CConst(Of String)
			Return New CConst(Of String)(value, Microsoft.CodeAnalysis.SpecialType.System_String)
		End Function

		Friend Shared Function Create(ByVal value As Single) As CConst(Of Single)
			Return New CConst(Of Single)(value, Microsoft.CodeAnalysis.SpecialType.System_Single)
		End Function

		Friend Shared Function Create(ByVal value As Double) As CConst(Of Double)
			Return New CConst(Of Double)(value, Microsoft.CodeAnalysis.SpecialType.System_Double)
		End Function

		Friend Shared Function Create(ByVal value As DateTime) As CConst(Of DateTime)
			Return New CConst(Of DateTime)(value, Microsoft.CodeAnalysis.SpecialType.System_DateTime)
		End Function

		Friend Shared Function CreateChecked(ByVal value As Object) As CConst
			Return CConst.TryCreate(RuntimeHelpers.GetObjectValue(value))
		End Function

		Friend Shared Function CreateNothing() As CConst(Of Object)
			Return New CConst(Of Object)(Nothing, Microsoft.CodeAnalysis.SpecialType.System_Object)
		End Function

		Friend Shared Function TryCreate(ByVal value As Object) As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst
			Dim cConst As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst
			If (value IsNot Nothing) Then
				Select Case SpecialTypeExtensions.FromRuntimeTypeOfLiteralValue(RuntimeHelpers.GetObjectValue(value))
					Case Microsoft.CodeAnalysis.SpecialType.System_Boolean
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToBoolean(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_Char
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToChar(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_SByte
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToSByte(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_Byte
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToByte(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_Int16
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToInt16(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_UInt16
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToUInt16(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_Int32
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToInt32(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_UInt32
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToUInt32(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_Int64
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToInt64(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_UInt64
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToUInt64(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_Decimal
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToDecimal(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_Single
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToSingle(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_Double
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToDouble(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_String
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToString(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_IntPtr
					Case Microsoft.CodeAnalysis.SpecialType.System_UIntPtr
					Case Microsoft.CodeAnalysis.SpecialType.System_Array
					Case Microsoft.CodeAnalysis.SpecialType.System_Collections_IEnumerable
					Case Microsoft.CodeAnalysis.SpecialType.System_Collections_Generic_IEnumerable_T
					Case Microsoft.CodeAnalysis.SpecialType.System_Collections_Generic_IList_T
					Case Microsoft.CodeAnalysis.SpecialType.System_Collections_Generic_ICollection_T
					Case Microsoft.CodeAnalysis.SpecialType.System_Collections_IEnumerator
					Case Microsoft.CodeAnalysis.SpecialType.System_Collections_Generic_IEnumerator_T
					Case Microsoft.CodeAnalysis.SpecialType.System_Collections_Generic_IReadOnlyList_T
					Case Microsoft.CodeAnalysis.SpecialType.System_Collections_Generic_IReadOnlyCollection_T
					Case Microsoft.CodeAnalysis.SpecialType.System_Nullable_T
					Label0:
						cConst = Nothing
						Exit Select
					Case Microsoft.CodeAnalysis.SpecialType.System_DateTime
						cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.Create(Convert.ToDateTime(RuntimeHelpers.GetObjectValue(value)))
						Exit Select
					Case Else
						GoTo Label0
				End Select
			Else
				cConst = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.CConst.CreateNothing()
			End If
			Return cConst
		End Function

		Public MustOverride Function WithError(ByVal id As ERRID) As CConst
	End Class
End Namespace